lib/open_feature/sdk/configuration.rb

```diff
@@ -14,76 +18,180 @@ class Configuration
       extend Forwardable
 
       attr_accessor :evaluation_context, :hooks
+      attr_reader :logger
 
       def initialize
         @hooks = []
         @providers = {}
         @provider_mutex = Mutex.new
+        @logger = nil  # Users can set a logger if needed
+        @event_emitter = EventEmitter.new(@logger)
+        @provider_state_registry = ProviderStateRegistry.new
       end
 
       def provider(domain: nil)
         @providers[domain] || @providers[nil]
       end
 
-      # When switching providers, there are a few lifecycle methods that need to be taken care of.
-      #   1. If a provider is already set, we need to call `shutdown` on it.
-      #   2. On the new provider, call `init`.
-      #   3. Finally, set the internal provider to the new provider
+      def logger=(new_logger)
+        @logger = new_logger
+        @event_emitter.logger = new_logger if @event_emitter
+      end
+
+      def add_handler(event_type, handler)
+        @event_emitter.add_handler(event_type, handler)
+      end
+
+      def remove_handler(event_type, handler)
+        @event_emitter.remove_handler(event_type, handler)
+      end
+
+      def clear_all_handlers
+        @event_emitter.clear_all_handlers
+      end
+
       def set_provider(provider, domain: nil)
+        old_provider = nil
+        
         @provider_mutex.synchronize do
-          @providers[domain].shutdown if @providers[domain].respond_to?(:shutdown)
-          provider.init if provider.respond_to?(:init)
+          old_provider = @providers[domain]
+          
+          begin
+            old_provider.shutdown if old_provider&.respond_to?(:shutdown)
+          rescue StandardError => e
+            @logger&.warn("Error shutting down previous provider #{old_provider&.class&.name || 'unknown'}: #{e.message}")
+          end
+          
           new_providers = @providers.dup
           new_providers[domain] = provider
           @providers = new_providers
+          
+          @provider_state_registry.set_initial_state(provider)
+          
+          if provider.is_a?(Provider::EventHandler)
+            provider.attach(ProviderEventDispatcher.new(self))
+          end
+          
+          Thread.new do
+            begin
+              if provider.respond_to?(:init)
+                begin
+                  provider.init(@evaluation_context)
+                rescue ArgumentError => e
+                  # Only fallback to no-args if it's specifically a "wrong number" error for 0-arity methods
```

Comment:
See change in `lib/open_feature/sdk/provider/in_memory_provider.rb`

This fallback continues support for
```
def init
  # legacy provider code
end
```

But other OpenFeature SDKs seem to expect that `evaluation_context` is always provided, i.e.
```
def init(evaluation_context = nil)
  # modern provider code
end
```